Add Husarion ROSbot XL and Husarion World

delihus opened the pull request:
**Description**
Fixed RPLidar A2 for mapping and navigation demos. Unify webots ROSbot 2R PROTO description with URDF description in [rosbot_ros](https://www.github.com/husarion/rosbot_ros). Added ROSbot XL and Husarion World to webots

**Related  Issues**
This pull-request is related to these issues #5540, https://github.com/cyberbotics/webots_ros2/issues/635. They describes bug with Lidars and slam-toolbox

**Tasks**
Add the list of tasks of this PR.
  - [x] create demo with ROSbot 2 and navigation2 https://github.com/husarion/rosbot-navigation
  - [x] add husarion world
  - [x] add ROSbot XL 

**Documentation**
Webots docker demo is described in the [webots-docker](https://github.com/husarion/webots-docker). 
Dockerized webots demo with slam-toolbox is the [rosbot-mapping](https://github.com/husarion/rosbot-mapping).
Dockerized webots demo with nav2 is in the [rosbot-navigation](https://github.com/husarion/rosbot-navigation)

**Screenshots**
![image](https://user-images.githubusercontent.com/109142865/223724276-f832f8d5-b98c-4d00-80f8-a006a2970876.png)
![image](https://user-images.githubusercontent.com/109142865/223724334-32070fa5-b90a-450e-96fb-15b5c3d2feb4.png)

omichel commented:
It seems this branch is out-of-date with the base branch. Could you please synchronize it with the upstream?

ad-daniel commented:
Hi, thank you for the contribution! It seems that in the process of rebasing the pull request to target the develop branch, some files have been changed that should not (basically all references to R2023a should be R2023b). There's also a couple of issues with missing icon for the added PROTO (you can use this utility: https://github.com/cyberbotics/webots/wiki/Adding-a-New-Robot#icon-studio to create one) and some wrong headers (again, should reference R2023b).
If you user icon studio for the icon, I suggest you remove all the other robots from the [json](https://github.com/cyberbotics/webots/blob/master/scripts/icon_studio/controllers/icon_creator/appearances.json) so that you just create the icon for your own. I'll take a closer look at the PR when the tests pass

ad-daniel commented on projects/robots/husarion/rosbot/controllers/rosbot/rosbot.c:
```diff
@@ -73,10 +73,10 @@ int main(int argc, char *argv[]) {
-  rear_left_position_sensor = wb_robot_get_device("rear left wheel motor sensor");
-  rear_right_position_sensor = wb_robot_get_device("rear right wheel motor sensor");
+  front_left_position_sensor = wb_robot_get_device("fl_range");
+  front_right_position_sensor = wb_robot_get_device("fr_range");
```
Any reason for the name change? From the new name they might be misjudged for range finders where in fact they are position sensors. Also keep in mind that sensors names will show up in the webots.cloud page (https://webots.cloud/run?version=R2023b&url=https://github.com/cyberbotics/webots/blob/develop/projects/robots/husarion/rosbot/protos/Rosbot.proto&type=undefined) in the "Devices" tab, as such complete names are preferable.

delihus replied:
Ohhh... Your right it shouldn't be changed. I mistook a wheel position sensor with a distance sensor.

ad-daniel commented on projects/robots/husarion/rosbot/controllers/rosbot/rosbot.c:
```diff
@@ -102,10 +102,10 @@ int main(int argc, char *argv[]) {
-  distance_sensors[1] = wb_robot_get_device("front left distance sensor");
-  distance_sensors[2] = wb_robot_get_device("front right distance sensor");
-  distance_sensors[3] = wb_robot_get_device("rear right distance sensor");
+  distance_sensors[0] = wb_robot_get_device("fl_range");
+  distance_sensors[1] = wb_robot_get_device("rl_range");
+  distance_sensors[2] = wb_robot_get_device("fr_range");
+  distance_sensors[3] = wb_robot_get_device("rr_range");
```
same thing, also I suspect we're introducing a bug here since two different devices are being given the same name (both distance sensors and position sensors). I suggest keep the previous nomenclature for all (motor & devices)

ad-daniel replied:
I see, but they're not very descriptive (like in the case of calling a Lidar "laser"), especially since in Webots we have other devices with similar names that could create confusion. Is it not possible perhaps to improve the URDF itself with more suitable names? Or maybe to just do the change on the webots side and remap the frames in ROS? I understand the change, but to be honest it doesn't seem very clean. Or at least, it seems like a step back compared to the previous PROTO

ad-daniel commented on projects/robots/husarion/rosbot/protos/Rosbot.proto:
```diff
@@ -717,4 +716,4 @@ PROTO Rosbot [
     }
   }
-}
+}
```
```suggestion
}

```

ad-daniel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,253 @@
+  field  SFBool      synchronization TRUE        # Is `Robot.synchronization`.
+  field  SFBool      selfCollision   FALSE       # Is `Robot.selfCollision`.
+  field  MFNode      lidarSlot       [ RpLidarA2 { name "laser" } ] # Extends the robot with a RpLidar A2 lidar.
+
```
```suggestion
```

ad-daniel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,253 @@
+      }
+    ]
+  }
+}
```
```suggestion
}

```

ad-daniel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,253 @@
+  field  SFBool      supervisor      FALSE       # Is `Robot.supervisor`.
+  field  SFBool      synchronization TRUE        # Is `Robot.synchronization`.
+  field  SFBool      selfCollision   FALSE       # Is `Robot.selfCollision`.
+  field  MFNode      lidarSlot       [ RpLidarA2 { name "laser" } ] # Extends the robot with a RpLidar A2 lidar.
```
same applies to this lidar

delihus replied:
Same case here we, use in [ros_components_description](https://github.com/husarion/ros_components_description/blob/ros2/urdf/slamtec_rplidar_a2.urdf.xacro) we use `laser` as a frame name of scan.

ad-daniel reviewed (changes requested):
- body_colision.stl has a typo

ad-daniel commented:
ops, submitted by mistake, still looking into it, give me a couple of more minutes

ad-daniel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,253 @@
+        }
+        device [
+          RotationalMotor {
+            name "rr_wheel_joint"
```
same comments in this PROTO about device names

delihus replied:
Also here the name is the same as in [rosbot_xl_description](https://github.com/husarion/rosbot_xl_ros/blob/master/rosbot_xl_description/urdf/rosbot_xl_macro.urdf.xacro#L74)

ad-daniel commented on projects/objects/husarion_world/protos/HusarionWorld.proto:
```diff
@@ -0,0 +1,44 @@
+# license: Apache License 2.0
+# license url: http://www.apache.org/licenses/LICENSE-2.0
+# World with Husarion Logo
+
```
EDIT: see other comment, it might make more sense to have this as a world rather than a PROTO

delihus commented on projects/devices/tdk/protos/Mpu-9250.proto:
```diff
@@ -34,6 +34,9 @@ PROTO Mpu-9250 [
       Compass {
         name %<= '"' + fields.name.value + ' compass"' >%
       }
+      InertialUnit {
+        name %<= '"' + fields.name.value + ' inertial_unit"' >%
+      }
```
We use in [rosbot_ros2_firmware](https://github.com/husarion/rosbot_ros2_firmware) Mpu-9250 with give us the orientation of the robot.
The best way to add to the ROSbots the IMU would be to implement  `webots_imu_hardware_interface' with `ros2_control` and then use `imu_broadcaster` from `ros2_controllers`.

ad-daniel replied:
I see, but since in webots this device gives ground truth information, is not very realistic. In webots_ros2 the IMU is also [modeled](https://github.com/cyberbotics/webots_ros2/wiki/References-Devices#imu-device) using the three devices (gyro, accelerometer & compass). Perhaps a middle ground could be to add a new boolean parameter to the PROTO that adds/removes the presence of this inertial unit, that gives the users the choice to either use the ground truth information or the more realistic approach. This can be achieved easily with [javascript templating](https://cyberbotics.com/doc/reference/javascript-procedural-proto) in the PROTO. You just need to add `# template language: javascript` to the header of the PROTO and wrap the device:
```

...

field SFBool includeInertialUnit FALSE

...

%< if (fields.includeInertialUnit.value) { >%
      InertialUnit {
        name %<= '"' + fields.name.value + ' inertial_unit"' >%
      }
%< } >%
```

delihus replied:
@ad-daniel 
Maybe could I add Inertial Unit to the Rosbot.proto and RosbotXl.proto?

ad-daniel replied:
In theory yes, but I think the other approach is better since the `Mpu-9250` is what is available in the real robot. That way it's up to the user then which to use, whether the ground truth provided by the inertial unit or the sensor fused information.

delihus commented:
> HusarionWorld was made as a PROTO, but probably makes more sense to have it as a world file (.wbt), this world could be used to showcase the RosbotXl robot in it, since currently doesn't appear anywhere. Basically to do something similar to what was done in the `rosbot.wbt` world to showcase the rosbot itself.
> 
> Also, the controller for the rosbotXl (named rosbot_xl) doesn't appear to be included

Where could I put the mesh of the Husarion World?

ad-daniel commented:
> Where could I put the mesh of the Husarion World?

You can create a meshes folder inside the worlds folder

delihus commented:
I tried to copy the Solid tag to the .wbt file and I got a segmentation fault.
```
 Solid {
        translation IS translation
        rotation IS rotation
        name IS name
        scale 10 10 10
        children [
            DEF BORDER Shape {
                geometry Mesh {
                    url "meshes/husarion_world.dae"
                    materialIndex 0
                }
                appearance PBRAppearance {
                    baseColor 0.6307572 0.005605392 0.02415763
                }
            }
            DEF LOGO Shape {
                geometry Mesh {
                    url "meshes/husarion_world.dae"
                    materialIndex 1
                }
                appearance PBRAppearance {
                    baseColor 1 1 1
                }
            }
        ]
        boundingObject Group{
            children[
                USE LOGO
                USE BORDER
            ]
        }
    }
 ```
 
 What is the best way to add .dae with boundingObject to the world file @ad-daniel?

ad-daniel commented:
The way the node is defined is correct, the issue is that the IS statement is exclusive to PROTO only, it cannot be used in worlds. Here's a complete working example based on your code, that you can modify at will:
```
#VRML_SIM R2023b utf8

EXTERNPROTO "webots://projects/objects/backgrounds/protos/TexturedBackground.proto"
EXTERNPROTO "webots://projects/objects/backgrounds/protos/TexturedBackgroundLight.proto"
EXTERNPROTO "webots://projects/objects/floors/protos/RectangleArena.proto"

WorldInfo {
}
Viewpoint {
}
TexturedBackground {
}
TexturedBackgroundLight {
}
Solid {
  scale 10 10 10
  children [
    DEF BORDER Shape {
      geometry Mesh {
        url "meshes/husarion_world.dae"
        materialIndex 0
      }
      appearance PBRAppearance {
        baseColor 0.6307572 0.005605392 0.02415763
      }
    }
    DEF LOGO Shape {
      geometry Mesh {
        url "meshes/husarion_world.dae"
        materialIndex 1
      }
      appearance PBRAppearance {
        baseColor 1 1 1
      }
    }
  ]
  boundingObject Group{
    children[
      USE LOGO
      USE BORDER
    ]
  }
}
```


Either way it shouldn't cause a segmentation fault..

delihus commented:
I moved Husarion World to the ROSbot XL folder and added `includeInertialUnit` flag to Mpu-9250.

@ad-daniel Any conclusions how to work with these joint names and lidar names?

The only I changed these names is to unify Webots names and our URDFs.

ad-daniel commented on projects/robots/husarion/rosbot_xl/worlds/rosbot_xl.wbt:
```diff
@@ -0,0 +1,73 @@
+#VRML_SIM R2023a utf8
```
```suggestion
#VRML_SIM R2023b utf8
```

ad-daniel reviewed (changes requested):
Thank you

rosbot.wbt is currently broken when run, the name of the lidar needs to be adjusted in the controller as well, not just the PROTO.

> Any conclusions how to work with these joint names and lidar names?

As mentioned [here](https://github.com/cyberbotics/webots/pull/5973#discussion_r1132059255), it would be better to have more descriptive names, but if there's no way around it then we can go ahead with the new names you propose (same as URDF)

delihus commented:
> Thank you
> 
>     * rosbot.wbt is currently broken when run, the name of the lidar needs to be adjusted in the controller as well, not just the PROTO.
> 
> 
> > Any conclusions how to work with these joint names and lidar names?
> 
> As mentioned [here](https://github.com/cyberbotics/webots/pull/5973#discussion_r1132059255), it would be better to have more descriptive names, but if there's no way around it then we can go ahead with the new names you propose (same as URDF)

The main purpose of changing the lidar name is a separated robot description of ROSbots.
For ROSbot 2R I would like to have `<webots>` tag describing ros2_control and imu plugins in [rosbot_ros](https://github.com/husarion/rosbot_ros/blob/humble/rosbot_description/urdf/rosbot_macro.urdf.xacro#L138) and the other `<webots>` tag describing lidar where I could change `frameName` in [ros_components_description](https://github.com/husarion/ros_components_description/blob/ros2/urdf/slamtec_rplidar_a2.urdf.xacro). Unlucky as I tested 2 separated `<webots>` tags only the one was loaded to `webots_ros2_driver`.

If it would be possible to use multiple `<webots>` tags in URDF I can change names in .proto to the `lidar`.

On the other hand default [slamtec configuration  ](https://github.com/husarion/sllidar_ros2) for RPLidar gives `laser` frame not `lidar` but this is the ROS package.

ad-daniel commented:
I don't think you can use multiple <webots> tags (at the moment), so the name changes you made are fine. 
The last remaining issue is that when running `rosbot.wbt` an error is thrown because of the name change

delihus commented:
> I don't think you can use multiple tags (at the moment), so the name changes you made are fine. The last remaining issue is that when running `rosbot.wbt` an error is thrown because of the name change

Do you mean this `lidar` name I changed yes?

ad-daniel commented:
Yes, I think you just did the change. The PR looks good to me, please also add an entry to the [changelog](https://github.com/cyberbotics/webots/blob/develop/docs/reference/changelog-r2023.md) (following the same format as the others) about the newly added rosbot-xl robot (under new feature), and set the pull request to ready for review, so that I can accept it.
Thank you!

ad-daniel commented on docs/reference/changelog-r2023.md:
```diff
@@ -5,6 +5,7 @@ Released on ??
   - New Features
     - Added a new launcher to simplify the start of extern controllers ([#5629](https://github.com/cyberbotics/webots/pull/5629)).
     - Added a warning when Webots and controller library versions are different ([#5896](https://github.com/cyberbotics/webots/pull/5896)).
+    - Added a model of the the ROSbot XL and the Husarion World from [Husarion](https://husarion.com/) and the example world ([#5973](https://github.com/cyberbotics/webots/pull/5973)).
```
```suggestion
    - Added model of the [Husarion](https://husarion.com/)'s ROSbot XL robot and a sample world ([#5973](https://github.com/cyberbotics/webots/pull/5973)).
```

ad-daniel reviewed (changes requested):
Last details

ad-daniel reviewed (approved):
Thank you!

ygoumaz commented on projects/devices/slamtec/protos/RpLidarA2.proto:
```diff
@@ -58,7 +58,7 @@ PROTO RpLidarA2 [
     }
-    fieldOfView 1.57
+    fieldOfView 6.283184
     horizontalResolution 800
```
```suggestion
    horizontalResolution 400
```

The resolution is 0.9°. So if the field of view is 360° (6.28 rad), it should be 400 here.

delihus replied:
@ygoumaz I would like to have the same frequency define in .proto file and LaserScan topic in ROS.

I found the bug when the `updateRate` is not define the LaserScan is published with 30Hz but the period should be 0 [here](https://github.com/cyberbotics/webots_ros2/blob/master/webots_ros2_driver/src/plugins/Ros2SensorPlugin.cpp#L26).

I modified Ros2Lidar to unify .proto and ROS frequency. https://github.com/cyberbotics/webots_ros2/pull/674 
The `defaultFrequency 12` gives:
```
average rate: 10.266
        min: 0.087s max: 0.106s std dev: 0.00246s window: 5611
```
using `ros2 topic hz /scan`

ad-daniel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+          rotation 0.000000 0.000000 1.000000 0.000000
+          children [
+            Mesh {
+              url "meshes/body_collision.stl"
```
In Webots it's not considered good practice to use a mesh as boundingObject, as it's heavy to compute collisions and less reliable. Ideally it would be better to approximate the shape of the body of the robot using primitive shapes as done in the rosbot proto.

delihus replied:
I understand this but we used the same concept in ignition-gazebo to exactly check the collision with the manipulator in MoveIt 2.
Maybe in the feature we would like to add the open manipulator on ROSbot XL in webots.

ad-daniel replied:
It was just a suggestion to let you know what we consider "good practice", up to you

omichel replied:
It is often possible to closely approximate the actual bounding object with a composite bounding object made of a several basic primitives (boxes, spheres, cylinders, capsules) arranged together. This usually yields to a much better performance and better stability than a mesh.

delihus replied:
Should I try to change it or left the collision as a mesh?

delihus replied:
Okay so at this time I left this. I will change it when there are problems.

ad-daniel reviewed (changes requested):
A couple more details, sorry

delihus commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+  field  SFVec3f     translation     0 0 0
+  field  SFRotation  rotation        0 0 1 0
+  field  SFString    name            "rosbot_xl" # Is `Robot.name`.
+  field  SFString    controller      "rosbot_xl" # Is `Robot.controller`.
```
Oh... okay. I just copied from ROSbot but now as I understand the controller for ROSbot is declared in `rosbot/controllers` but for ROSbot XL is not implemented. Am I right?

ad-daniel replied:
yes exactly

ad-daniel reviewed (dismissed):
Fine from my side

delihus commented:
Are there any requested changes or we are waiting for @ygoumaz review?

ygoumaz reviewed (approved):
All good to me!

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+    selfCollision IS selfCollision
+    physics Physics {
+      density -1
+      mass 3.500000
```
```suggestion
      mass 3.5
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+    boundingObject DEF body_collision Group {
+      children [
+        DEF MAIN_BODY_BOX Transform {
+          translation 0.000000 0.000000 0.048
+          rotation 0.000000 0.000000 1.000000 0.000000
```
```suggestion
          translation 0 0 0.048
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+    }
+    children [
+      DEF MAIN_BODY Transform {
+        translation 0.000000 0.000000 0.048000
+        rotation 0.000000 0.000000 1.000000 0.000000
```
```suggestion
        translation 0 0 0.048
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+
+      HingeJoint {
+        jointParameters HingeJointParameters {
+          axis 0.000000 1.000000 0.000000
+          anchor 0.085000 0.135000 0.048000
```
```suggestion
          axis 0 1 0
          anchor 0.085 0.135 0.048
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+        device [
+          RotationalMotor {
+            name "fl_wheel_joint"
+            maxVelocity 26.0
+            maxTorque 100.0
```
```suggestion
            maxVelocity 26
            maxTorque 100
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+        jointParameters HingeJointParameters {
+          axis 0.000000 1.000000 0.000000
+          anchor 0.085000 0.135000 0.048000
+          dampingConstant 5.0
+          staticFriction 1.0
```
```suggestion
          dampingConstant 5
          staticFriction 1
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+          }
+        ]
+        endPoint Solid {
+          translation 0.085000 0.135000 0.048000
+          rotation 0.000000 0.000000 1.000000 0.000000
```
```suggestion
          translation 0.085 0.135 0.048
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+          rotation 0.000000 0.000000 1.000000 0.000000
+          children [
+            Transform {
+              translation 0.000000 0.000000 0.000000
+              rotation 0.000000 0.000000 1.000000 0.000000
```
```suggestion
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+          ]
+          name "front left wheel"
+          boundingObject DEF LEFT_WHEEL_CYLINDER Transform {
+            translation 0.000000 0.000000 0.000000
+            rotation 1.000000 0.000000 0.000000 -1.570800
```
```suggestion
            rotation 1 0 0 -1.5708
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+          physics DEF WHEEL_PHYSICS Physics {
+            density -1
+            mass 0.051000
+            centerOfMass [ 0.000000 0.000000 0.000000 ]
```
```suggestion
            centerOfMass [ 0 0 0 ]
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+            centerOfMass [ 0.000000 0.000000 0.000000 ]
+            inertiaMatrix [
+              3.954000e-05 7.116000e-05 4.032000e-05
+              0.000000e+00 0.000000e+00 0.000000e+00
```
```suggestion
              0 0 0
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+      }
+      HingeJoint {
+        jointParameters HingeJointParameters {
+          axis 0.000000 1.000000 0.000000
+          anchor 0.085000 -0.135000 0.048000
+          dampingConstant 5.0
+          staticFriction 1.0
```
```suggestion
          axis 0 1 0
          anchor 0.085 -0.135 0.048
          dampingConstant 5
          staticFriction 1
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+        device [
+          RotationalMotor {
+            name "fr_wheel_joint"
+            maxVelocity 26.0
+            maxTorque 100.0
```
```suggestion
            maxVelocity 26
            maxTorque 100
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+          }
+        ]
+        endPoint Solid {
+          translation 0.085000 -0.135000 0.048000
+          rotation 0.000000 0.000000 1.000000 0.000000
```
```suggestion
          translation 0.085 -0.135 0.048
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+          rotation 0.000000 0.000000 1.000000 0.000000
+          children [
+            Transform {
+              translation 0.000000 0.000000 0.000000
+              rotation 0.000000 0.000000 1.000000 3.141592
```
```suggestion
              rotation 0 0 1 3.141592
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+          ]
+          name "front right wheel"
+          boundingObject DEF RIGHT_WHEEL_CYLINDER Transform {
+            translation 0.000000 0.000000 0.000000
+            rotation 1.000000 0.000000 0.000000 -1.570800
```
```suggestion
            rotation 1 0 0 -1.5708
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+
+      HingeJoint {
+        jointParameters HingeJointParameters {
+          axis 0.000000 1.000000 0.000000
+          anchor -0.085000 0.135000 0.048000
+          dampingConstant 5.0
+          staticFriction 1.0
```
```suggestion
          axis 0 1 0
          anchor -0.085 0.135 0.048
          dampingConstant 5
          staticFriction 1
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+        device [
+          RotationalMotor {
+            name "rl_wheel_joint"
+            maxVelocity 26.0
+            maxTorque 100.0
```
```suggestion
            maxVelocity 26
            maxTorque 100
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+          }
+        ]
+        endPoint Solid {
+          translation -0.085000 0.135000 0.048000
+          rotation 0.000000 0.000000 1.000000 0.000000
```
```suggestion
          translation -0.085 0.135 0.048
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+          translation -0.085000 0.135000 0.048000
+          rotation 0.000000 0.000000 1.000000 0.000000
+          children [
+            Transform {
+              translation 0.000000 0.000000 0.000000
+              rotation 0.000000 0.000000 1.000000 0.000000
+              children [
+                USE WHEEL_SHAPE_A
+              ]
+            }
```
```suggestion
            USE WHEEL_SHAPE_A
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+          ]
+          name "rear left wheel"
+          boundingObject DEF LEFT_WHEEL_CYLINDER Transform {
+            translation 0.000000 0.000000 0.000000
+            rotation 1.000000 0.000000 0.000000 -1.570800
```
```suggestion
            rotation 1 0 0 -1.5708
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+
+      HingeJoint {
+        jointParameters HingeJointParameters {
+          axis 0.000000 1.000000 0.000000
+          anchor -0.085000 -0.135000 0.048000
+          dampingConstant 5.0
+          staticFriction 1.0
```
```suggestion
          axis 0 1 0
          anchor -0.085 -0.135 0.048
          dampingConstant 5
          staticFriction 1
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+        device [
+          RotationalMotor {
+            name "rr_wheel_joint"
+            maxVelocity 26.0
+            maxTorque 100.0
```
```suggestion
            maxVelocity 26
            maxTorque 100
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+          }
+        ]
+        endPoint Solid {
+          translation -0.085000 -0.135000 0.048000
+          rotation 0.000000 0.000000 1.000000 0.000000
```
```suggestion
          translation -0.085 -0.135 0.048
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+          rotation 0.000000 0.000000 1.000000 0.000000
+          children [
+            Transform {
+              translation 0.000000 0.000000 0.000000
+              rotation 0.000000 0.000000 1.000000 3.141592
```
```suggestion
              rotation 0 0 1 3.141592
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+          ]
+          name "rear right wheel"
+          boundingObject DEF RIGHT_WHEEL_CYLINDER Transform {
+            translation 0.000000 0.000000 0.000000
+            rotation 1.000000 0.000000 0.000000 -1.570800
```
```suggestion
            rotation 1 0 0 -1.5708
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+        }
+      }
+      DEF LIDAR_SLOT Transform {
+        translation 0.000000 0.000000 0.13145
```
```suggestion
        translation 0 0 0.13145
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,254 @@
+        children IS lidarSlot
+      }
+      Mpu-9250 {
+        translation -0.136900 -0.041900 0.037000
+        rotation 1.000000 0.000000 0.000000 3.14159
```
```suggestion
        translation -0.1369 -0.0419 0.037
        rotation 1 0 0 3.14159
```

omichel commented on projects/robots/husarion/rosbot_xl/protos/RosbotXl.proto:
```diff
@@ -0,0 +1,238 @@
+        endPoint Solid {
+          translation 0.085 0.135 0.048
+          children [
+            Transform {
+              children [
+                DEF WHEEL_SHAPE_B CadShape {
+                  url "meshes/mecanum_b.dae"
+                }
+              ]
+            }
```
```suggestion
            DEF WHEEL_SHAPE_B CadShape {
              url "meshes/mecanum_b.dae"
            }
```

ad-daniel commented:
@delihus could you synchronize the branch so that this PR can be merged? Thank you

delihus commented:
synced

ad-daniel commented:
Thank you!

delihus commented:
Thank you!
I just requested review for `webots_ros2` https://github.com/cyberbotics/webots_ros2/pull/682.
